File: banking-component/src/test/java/de/rwth/swc/banking/BankingControllerTest.java
package de.rwth.swc.banking;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import de.rwth.swc.interact.junit.jupiter.annotation.InterACtTest;
import de.rwth.swc.interact.junit.jupiter.annotation.Offset;
import de.rwth.swc.interact.rest.RestMessage;
import kotlin.Unit;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.aggregator.AggregateWith;
import org.junit.jupiter.params.provider.CsvSource;
import org.mockserver.integration.ClientAndServer;
import org.mockserver.matchers.Times;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.context.annotation.Import;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;

import static de.rwth.swc.interact.test.ExampleBasedAssertionsKt.forExample;
import static de.rwth.swc.interact.test.PropertyBasedAssertionsKt.inherently;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockserver.model.HttpRequest.request;
import static org.mockserver.model.HttpResponse.response;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Import({TestConfig.class})
class BankingControllerTest {

    @LocalServerPort
    private int port;
    private ObjectMapper mapper = new ObjectMapper();
    private BankingApi bankingApi;
    private static ClientAndServer mockServer;

    @BeforeAll
    public static void init() {
        mockServer = ClientAndServer.startClientAndServer(8082);
    }

    @AfterAll
    public static void stop() {
        mockServer.stop();
    }

    @BeforeEach
    public void setUp() {
        bankingApi = TestConfig.bankingApi(port);
        mockServer.reset();
    }

    @InterACtTest
    @CsvSource({"300, DE19 5001 0517 5326 8513 68, GE13617195993119486971, false"})
    public void v1WhenTransferIsSendAndIbanValidationFailsItFails(
            @AggregateWith(TransferAggregator.class) RestMessage<Transfer> transfer,
            @Offset(3) @AggregateWith(BooleanAggregator.class) RestMessage<Boolean> ibanValidatorResponse
    ) {

        mockServer.when(
                request().withMethod(HttpMethod.POST.name())
                        .withPath("/v1/validate/iban")
                        .withBody(transfer.getBody().fromIban),
                Times.exactly(1)
        ).respond(
                response().withStatusCode(HttpStatus.OK.value())
                        .withContentType(org.mockserver.model.MediaType.APPLICATION_JSON)
                        .withBody(String.valueOf(ibanValidatorResponse.getBody()))
        );

        var result = bankingApi.sendMoneyV1(transfer.getBody());

        inherently(() -> {
            assertThat(result.getBody()).isEqualTo(false);
            assertThat(result.getStatusCode()).isEqualTo(HttpStatus.OK);
            return Unit.INSTANCE;
        });
    }

    @InterACtTest
    @CsvSource({
            "500, DE93 5001 0517 6966 2689 58, GE13617195993119486971, true, true, true",
            "200, DE93 5001 0517 6966 2689 58, DE19 5001 0517 5326 8513 68, true, false, false"})
    public void v1WhenTransferIsSendAndIbanValidationSucceedsItRespondsAccordingToAmountValidation(
            @AggregateWith(TransferAggregator.class) RestMessage<Transfer> transfer,
            @Offset(3) @AggregateWith(BooleanAggregator.class) RestMessage<Boolean> ibanValidatorResponse,
            @Offset(4) @AggregateWith(BooleanAggregator.class) RestMessage<Boolean> amountValidatorResponse,
            @Offset(5) Boolean expectedResponse) throws JsonProcessingException {

        mockServer.when(
                request().withMethod(HttpMethod.POST.name())
                        .withPath("/v1/validate/iban")
                        .withBody(transfer.getBody().fromIban),
                Times.exactly(1)
        ).respond(
                response().withStatusCode(HttpStatus.OK.value())
                        .withContentType(org.mockserver.model.MediaType.APPLICATION_JSON)
                        .withBody(String.valueOf(ibanValidatorResponse.getBody()))
        );

        mockServer.when(
                request().withMethod(HttpMethod.POST.name())
                        .withPath("/v1/validate/amount")
                        .withBody(mapper.writeValueAsString(transfer.getBody())),
                Times.exactly(1)
        ).respond(
                response().withStatusCode(HttpStatus.OK.value())
                        .withContentType(org.mockserver.model.MediaType.APPLICATION_JSON)
                        .withBody(String.valueOf(amountValidatorResponse.getBody()))
        );

        var result = bankingApi.sendMoneyV1(transfer.getBody());

        forExample(() -> {
            assertThat(result.getBody()).isEqualTo(expectedResponse);
            return Unit.INSTANCE;
        });
        inherently(() -> {
            assertThat(result.getStatusCode()).isEqualTo(HttpStatus.OK);
            return Unit.INSTANCE;
        });
    }

    @InterACtTest
    @CsvSource({
            "500, DE93 5001 0517 6966 2689 58, GE13617195993119486971, true, true",
            "200, GE13617195993119486971, DE19 5001 0517 5326 8513 68, false, false"})
    public void v2WhenTransferIsSendItIsHandled(
            @AggregateWith(TransferAggregator.class) RestMessage<Transfer> transfer,
            @Offset(3) @AggregateWith(BooleanAggregator.class) RestMessage<Boolean> validationResponse,
            @Offset(4) Boolean expectedResponse) throws JsonProcessingException {

        mockServer.when(
                request().withMethod(HttpMethod.POST.name())
                        .withPath("/v2/validate/iban")
                        .withBody(mapper.writeValueAsString(transfer.getBody())),
                Times.exactly(1)
        ).respond(
                response().withStatusCode(HttpStatus.OK.value())
                        .withContentType(org.mockserver.model.MediaType.APPLICATION_JSON)
                        .withBody(String.valueOf(validationResponse.getBody()))
        );

        var result = bankingApi.sendMoneyV2(transfer.getBody());

        forExample(() -> {
            assertThat(result.getBody()).isEqualTo(expectedResponse);
            return Unit.INSTANCE;
        });

        inherently(() -> {
            assertThat(result.getStatusCode()).isEqualTo(HttpStatus.OK);
            return Unit.INSTANCE;
        });
    }
}

File: banking-component/src/test/kotlin/de/rwth/swc/banking/BankingControllerTest.kt
package de.rwth.swc.banking


import com.fasterxml.jackson.databind.ObjectMapper
import de.rwth.swc.interact.domain.serialization.SerializationConstants
import de.rwth.swc.interact.junit.jupiter.annotation.InterACtTest
import de.rwth.swc.interact.junit.jupiter.annotation.Offset
import de.rwth.swc.interact.rest.RestMessage
import de.rwth.swc.interact.test.forExample
import de.rwth.swc.interact.test.inherently
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.aggregator.AggregateWith
import org.junit.jupiter.params.provider.CsvSource
import org.mockserver.integration.ClientAndServer
import org.mockserver.matchers.Times
import org.mockserver.model.HttpRequest
import org.mockserver.model.HttpResponse
import org.mockserver.model.MediaType
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.server.LocalServerPort
import org.springframework.context.annotation.Import
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.web.reactive.function.client.WebClient


@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Import(
    TestConfig::class
)
internal class BankingControllerTest {

    @LocalServerPort
    private var port: Int = 0

    @Autowired
    private lateinit var webClientBuilder: WebClient.Builder

    private lateinit var bankingApi: BankingApi

    @Autowired
    private lateinit var mapper: ObjectMapper

    @Autowired
    private lateinit var bankAccountStore: BankAccountStore

    @BeforeEach
    fun setUp() {
        mockServer.reset()
        bankingApi = TestConfig.bankingApi(webClientBuilder, port)
    }

    @Nested
    inner class ApiV1 {
        @InterACtTest
        @CsvSource("DE19 5001 0517 5326 8513 68, invalidIban, 300, false")
        fun `when IBAN validation fails the transfer fails`(
            @AggregateWith(TransferAggregator::class) transfer: RestMessage<Transfer>,
            @Offset(3) @AggregateWith(IbanValidationResponseAggregator::class) ibanValidatorResponse: RestMessage<IbanValidationResponse>
        ) {
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v1/validate/iban")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(ibanValidatorResponse.body))
            )
            val result = bankingApi.transferV1(transfer.body)
            inherently {
                assertThat(result.body).isInstanceOf(TransferResponse.Failure::class.java)
                assertThat((result.body as TransferResponse.Failure).reason)
                    .isEqualTo((ibanValidatorResponse.body as IbanValidationResponse.Invalid).errorMessage)
                assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
            }
        }

        @InterACtTest
        @CsvSource(
            "DE93 5001 0517 6966 2689 58, GE13617195993119486971, 500, true, true"
        )
        fun `when IBAN validation succeeds and IBAN is not on blacklist and the account has sufficient funds the transfer is successful`(
            @AggregateWith(TransferAggregator::class) transfer: RestMessage<Transfer>,
            @Offset(3) @AggregateWith(IbanValidationResponseAggregator::class) ibanValidatorResponse: RestMessage<IbanValidationResponse>,
            @Offset(4) @AggregateWith(BlacklistCheckResponseAggregator::class) blacklistCheckResponse: RestMessage<BlacklistCheckResponse>
        ) {
            bankAccountStore.addBankAccount(transfer.body.fromIban, transfer.body.amount)
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v1/validate/iban")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(ibanValidatorResponse.body))
            )
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v1/check/blacklist")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(blacklistCheckResponse.body))
            )
            val result = bankingApi.transferV1(transfer.body)
            inherently {
                assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
                assertThat(result.body).isInstanceOf(TransferResponse.Success::class.java)
                assertThat(bankAccountStore.getBalance(transfer.body.fromIban)).isEqualTo(0)
            }
            bankAccountStore.clear()
        }

        @InterACtTest
        @CsvSource(
            "DE93 5001 0517 6966 2689 58, GE13617195993119486971, 500, true, true"
        )
        fun `when IBAN validation succeeds and IBAN is not on blacklist but the account has insufficient funds the transfer fails`(
            @AggregateWith(TransferAggregator::class) transfer: RestMessage<Transfer>,
            @Offset(3) @AggregateWith(IbanValidationResponseAggregator::class) ibanValidatorResponse: RestMessage<IbanValidationResponse>,
            @Offset(4) @AggregateWith(BlacklistCheckResponseAggregator::class) blacklistCheckResponse: RestMessage<BlacklistCheckResponse>
        ) {
            bankAccountStore.addBankAccount(transfer.body.fromIban, transfer.body.amount - 1)
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v1/validate/iban")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(ibanValidatorResponse.body))
            )
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v1/check/blacklist")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(blacklistCheckResponse.body))
            )
            val result = bankingApi.transferV1(transfer.body)
            inherently {
                assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
                assertThat(result.body).isInstanceOf(TransferResponse.Failure::class.java)
                assertThat((result.body as TransferResponse.Failure).reason)
                    .isEqualTo("Insufficient funds")
            }
            bankAccountStore.clear()
        }

        @InterACtTest
        @CsvSource(
            "DE93 5001 0517 6966 2689 58, DE19 5001 0517 5326 8513 68, 200, true, false"
        )
        fun `when IBAN validation succeeds but IBAN is on blacklist the transfer fails`(
            @AggregateWith(TransferAggregator::class) transfer: RestMessage<Transfer>,
            @Offset(3) @AggregateWith(IbanValidationResponseAggregator::class) ibanValidatorResponse: RestMessage<IbanValidationResponse>,
            @Offset(4) @AggregateWith(BlacklistCheckResponseAggregator::class) blacklistCheckResponse: RestMessage<BlacklistCheckResponse>
        ) {
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v1/validate/iban")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(ibanValidatorResponse.body))
            )
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v1/check/blacklist")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(blacklistCheckResponse.body))
            )
            val result = bankingApi.transferV1(transfer.body)
            inherently {
                assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
                assertThat(result.body).isInstanceOf(TransferResponse.Failure::class.java)
                assertThat((result.body as TransferResponse.Failure).reason)
                    .isEqualTo((blacklistCheckResponse.body as BlacklistCheckResponse.Blacklisted).reason)
            }
        }
    }

    @Nested
    inner class ApiV2 {
        @InterACtTest
        @CsvSource(
            "DE93 5001 0517 6966 2689 58, GE13617195993119486971, 500, true"
        )
        fun `when IBAN validation succeeds but the account has insufficient funds the transfer fails`(
            @AggregateWith(TransferAggregator::class) transfer: RestMessage<Transfer>,
            @Offset(3) @AggregateWith(IbanValidationResponseAggregator::class) validationResponse: RestMessage<IbanValidationResponse>
        ) {
            bankAccountStore.addBankAccount(transfer.body.fromIban, transfer.body.amount - 1)
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v2/validate/iban")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(validationResponse.body))
            )
            val result = bankingApi.transferV2(transfer.body)
            inherently {
                assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
                assertThat(result.body).isInstanceOf(TransferResponse.Failure::class.java)
                assertThat((result.body as TransferResponse.Failure).reason)
                    .isEqualTo("Insufficient funds")
            }
            bankAccountStore.clear()
        }

        @InterACtTest
        @CsvSource(
            "DE19 5001 0517 5326 8513 68, invalidIban, 200, false"
        )
        fun `when IBAN validation fails the transfer should fail`(
            @AggregateWith(TransferAggregator::class) transfer: RestMessage<Transfer>,
            @Offset(3) @AggregateWith(IbanValidationResponseAggregator::class) validationResponse: RestMessage<IbanValidationResponse>
        ) {
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v2/validate/iban")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(validationResponse.body))
            )
            val result = bankingApi.transferV2(transfer.body)
            inherently {
                assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
                assertThat(result.body).isInstanceOf(TransferResponse.Failure::class.java)
                assertThat((result.body as TransferResponse.Failure).reason)
                    .isEqualTo((validationResponse.body as IbanValidationResponse.Invalid).errorMessage)
            }
        }

        @InterACtTest
        @CsvSource(
            "DE93 5001 0517 6966 2689 58, GE13617195993119486971, 500, true"
        )
        fun `when IBAN validation succeeds and the account has sufficient funds the transfer succeeds`(
            @AggregateWith(TransferAggregator::class) transfer: RestMessage<Transfer>,
            @Offset(3) @AggregateWith(IbanValidationResponseAggregator::class) validationResponse: RestMessage<IbanValidationResponse>
        ) {
            bankAccountStore.addBankAccount(transfer.body.fromIban, transfer.body.amount)
            mockServer.`when`(
                HttpRequest.request().withMethod(HttpMethod.POST.name())
                    .withPath("/api/v2/validate/iban")
                    .withBody(transfer.body.toIban),
                Times.exactly(1)
            ).respond(
                HttpResponse.response().withStatusCode(HttpStatus.OK.value())
                    .withContentType(MediaType.APPLICATION_JSON)
                    .withBody(mapper.writeValueAsString(validationResponse.body))
            )
            val result = bankingApi.transferV2(transfer.body)
            inherently {
                assertThat(result.statusCode).isEqualTo(HttpStatus.OK)
                assertThat(result.body).isInstanceOf(TransferResponse.Success::class.java)
                assertThat(bankAccountStore.getBalance(transfer.body.fromIban)).isEqualTo(0)
            }
            bankAccountStore.clear()
        }
    }

    companion object {

        private lateinit var mockServer: ClientAndServer

        @JvmStatic
        @BeforeAll
        fun init() {
            mockServer = ClientAndServer.startClientAndServer(8082)
        }

        @JvmStatic
        @AfterAll
        fun stop() {
            mockServer.stop()
        }
    }
}
